processing/src/main/java/org/apache/druid/segment/StringDimensionHandler.java

```diff
@@ -33,61 +33,44 @@
 
 public class StringDimensionHandler implements DimensionHandler<Integer, int[], String>
 {
-
   /**
-   * Compares {@link IndexedInts} lexicographically, with the exception that if a row contains only zeros (that's the
-   * index of null) at all positions, it is considered "null" as a whole and is "less" than any "non-null" row. Empty
-   * row (size is zero) is also considered "null".
-   *
-   * The implementation is a bit complicated because it tries to check each position of both rows only once.
+   * This comparator uses the following rules:
+   * - If the value arrays have different lengths, the shorter value array is considered smaller
+   *   - The single exception to this is null and the empty list, which are considered equal
+   * - If the value arrays are the same length, compare value by value until a difference is reached
    */
   private static final Comparator<ColumnValueSelector> DIMENSION_SELECTOR_COMPARATOR = (s1, s2) -> {
     IndexedInts row1 = getRow(s1);
     IndexedInts row2 = getRow(s2);
     int len1 = row1.size();
     int len2 = row2.size();
-    boolean row1IsNull = true;
-    boolean row2IsNull = true;
-    for (int i = 0; i < Math.min(len1, len2); i++) {
-      int v1 = row1.get(i);
-      row1IsNull &= v1 == 0;
-      int v2 = row2.get(i);
-      row2IsNull &= v2 == 0;
-      int valueDiff = Integer.compare(v1, v2);
-      if (valueDiff != 0) {
-        return valueDiff;
-      }
-    }
-    //noinspection SubtractionInCompareTo -- substraction is safe here, because lengths or rows are small numbers.
-    int lenDiff = len1 - len2;
-    if (lenDiff == 0) {
-      return 0;
-    } else {
-      if (!row1IsNull || !row2IsNull) {
-        return lenDiff;
-      } else {
-        return compareRestNulls(row1, len1, row2, len2);
+    int retVal = Integer.compare(len1, len2);
+    int valsIndex = 0;
+
+    if (retVal != 0) {
```

Comment:
Hmm, this sort seems to behave quite a bit differently than the intended function of the previous sort; this one ordering by row length then lexicographically element by element, while the previous ordering lexicographically element by element then by length.

```
["a", "b"],
["a", "c"],
["d", "e"],
["a", "b", "c"]
```
 compared to  
```
["a", "b"],
["a", "b", "c"],
["a", "c"],
["d", "e"]
```

It might be more prudent to adjust the other sort to be more like this one is intended to be, though I imagine in practice that most of the time this doesn't make much difference except when all the times are the same (so less query granularity) and this is the prominent ordering column, and even then might not make much difference in practice, so don't make this change probably if it is a major hassle and instead we just put in release notes that there is a behavior change.

Reply:
Good point, I've adjusted both comparators to follow the rules described here: https://github.com/apache/druid/pull/9742/files#diff-4c87c650e139d3889184df880b97ff4aR37